localstack/services/s3/virtual_host.py

```diff
@@ -79,6 +83,13 @@ def _rewrite_url(url: str, bucket: str, region: str) -> str:
         if region:
             netloc = netloc.replace(f"{region}", "")
 
+        # if the user specifies a custom hostname for LocalStack, this name may not be resolvable by
+        # LocalStack. We are proxying the request to ourself, so replace their custom hostname with
+        # `localhost.localstack.cloud` which also matches the PATH matchers.
+        host_definition = localstack_host(use_hostname_external=True)
+        if host_definition.host != LOCALHOST:
```

Comment:
I was confused by this at first, but it's because `HOSTNAME_EXTERNAL` is set to `LOCALHOST` if not set, is that right? Now it makes more sense to me.